packages/webamp/js/components/VisPainter.ts

```diff
@@ -0,0 +1,748 @@
+export interface Vis {
+  canvas?: HTMLCanvasElement;
+  colors: string[];
+  analyser?: AnalyserNode;
+  oscStyle?: "dots" | "solid" | "lines";
+  bandwidth?: "wide" | "thin";
+  coloring?: "fire" | "line" | "normal";
+  peaks?: boolean;
+  saFalloff?: "slower" | "slow" | "moderate" | "fast" | "faster";
+  saPeakFalloff?: "slower" | "slow" | "moderate" | "fast" | "faster";
+  sa?: "analyzer" | "oscilloscope" | "none";
+}
+import { FFT } from "./FFTNullsoft";
+import {
+  renderHeight,
+  smallVis,
+  PIXEL_DENSITY,
+  doubled,
+  isMWOpen,
+} from "./Vis";
+
+let saPeaks = new Int16Array(76).fill(0);
+let saData2 = new Float32Array(76).fill(0);
+let saData = new Int16Array(76).fill(0);
+let saFalloff = new Float32Array(76).fill(0);
+let sample = new Float32Array(76).fill(0);
+let barPeak = new Int16Array(76).fill(0); // Needs to be specified as Int16 else the peaks don't behave as they should
+let chunk: number;
+let uVar12: number;
+let falloff: number;
+let peakFalloff: number;
+
+let pushDown: number = 0;
+
+/**
+ * Base class of Visualizer (animation frame renderer engine)
+ */
+export class VisPaintHandler {
+  _vis: Vis;
+  _ctx: CanvasRenderingContext2D | null;
+
+  constructor(vis: Vis) {
+    this._vis = vis;
+    this._ctx = vis.canvas!.getContext("2d");
+  }
+
+  /**
+   * Attempt to build cached bitmaps for later use while rendering a frame.
+   * Purpose: fast rendering in animation loop
+   */
+  prepare() {}
+
+  /**
+   * Called once per frame rendering
+   */
+  paintFrame() {}
+
+  /**
+   * Attempt to cleanup cached bitmaps
+   */
+  dispose() {}
+}
+
+const SAMPLESIN = 1024;
+const SAMPLESOUT = 512;
+const inWaveData = new Float32Array(SAMPLESIN);
+const outSpectralData = new Float32Array(SAMPLESOUT);
```

Comment:
Same as above, I think the values in these arrays will likely get corrupted if there are multiple visualizers on the page at the same time.